src/archive/src/main.rs

```diff
@@ -0,0 +1,438 @@
+fn with_anchor_index_mut<R>(f: impl FnOnce(&mut AnchorIndex) -> R) -> R {
+    ANCHOR_INDEX.with(|cell| f(&mut *cell.borrow_mut()))
+}
+
+/// Configuration state of the archive.
+enum ConfigState {
+    Uninitialized, // This state is only used between wasm module initialization and init().
+    Initialized(ArchiveConfig),
+}
+
+impl ConfigState {
+    fn get(&self) -> &ArchiveConfig {
+        match &self {
+            ConfigState::Uninitialized => trap("archive config not initialized"),
+            ConfigState::Initialized(config) => config,
+        }
+    }
+}
+
+/// Configuration of the archive canister.
+#[derive(Clone, Debug, CandidType, Deserialize)]
+struct ArchiveConfig {
+    /// This canister will accept entries only from this principal.
+    ii_canister: Principal,
+    /// The maximum number of entries returned in a single read canister call.
+    max_entries_per_call: u16,
+    /// Timestamp of the last install / upgrade of this canister.
+    last_upgrade_timestamp: Timestamp,
+}
+
+impl Storable for ConfigState {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        match &self {
+            ConfigState::Uninitialized => Cow::Borrowed(&[]),
+            ConfigState::Initialized(config) => {
+                let buf = candid::encode_one(config).expect("failed to encode archive config");
+                Cow::Owned(buf)
+            }
+        }
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        if bytes.len() == 0 {
+            return ConfigState::Uninitialized;
+        }
+        ConfigState::Initialized(
+            candid::decode_one::<ArchiveConfig>(&bytes).expect("failed to decode archive config"),
+        )
+    }
+}
+
+/// Index key for the anchor index.
+/// Changing the (serialized) size of this value requires a stable memory migration.
+#[derive(Debug)]
+struct AnchorIndexKey {
+    anchor: Anchor,
+    timestamp: Timestamp,
+    log_index: u64,
+}
+
+impl Storable for AnchorIndexKey {
+    fn to_bytes(&self) -> Cow<[u8]> {
+        let mut buf = Vec::with_capacity(std::mem::size_of::<AnchorIndexKey>());
+        buf.extend(&self.anchor.to_le_bytes());
+        buf.extend(&self.timestamp.to_le_bytes());
+        buf.extend(&self.log_index.to_le_bytes());
+        Cow::Owned(buf)
+    }
+
+    fn from_bytes(bytes: Vec<u8>) -> Self {
+        AnchorIndexKey {
+            anchor: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[0..8]).expect("failed to read anchor"),
+            ),
+            timestamp: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[8..16]).expect("failed to read timestamp"),
+            ),
+            log_index: u64::from_le_bytes(
+                TryFrom::try_from(&bytes[16..]).expect("failed to read log_index"),
+            ),
+        }
+    }
+}
+
+#[update]
+fn write_entry(anchor: Anchor, timestamp: Timestamp, entry: ByteBuf) {
+    with_config(|config| {
+        if config.ii_canister != caller() {
```

Comment:
out of curiosity, is there an `api` call to figure out the controller of this canister?

Reply:
There is. One could call the `ManagementCanister` for the `canister_status` and extract it from there. But I think separating canister ownership from authorization to write entries is sensible, even if we keep them the same for now.

Reply:
I think that for simplicity's sake we may want to default to this when integrating with II (passing arguments is just another point of failure) although I'm guessing this'll stay necessary for testing, maybe?

Reply:
I just realized that in `init` and `post_upgrade` we cannot do a call to the management canister. So in order to find out the controller, we would have to delay initialization of the config into the first `update` call, which imho adds more complexity than simply passing in the argument.

I will leave this arg as it is for now and we can revisit when we do the integration with II.